### PR TITLE
Make dialogs scrollable when their content is larger than the window height

### PR DESCRIPTION
**Pull request by jameskoster**
This PR addresses some issues with the Dialog component. The main issue fixed is how modals that are larger than the current viewport are handled. Currently in these cases the edges of the modal bleed off-screen and are unaccessible. This can result in action buttons being un-clickable. 😱

I also removed the `color` declaration on `.dialog__content` so colors are now inherited as expected.

<hr />

To test: 

* Open up a dialog, for example the 'Add link' modal in the Editor. 
* Resize your browser window so that it is smaller than the Dialog.
* Notice the dialog content now scrolls and the action buttons are fixed in position.
* Check the media modal, (and any other modals) are working as expected.

<hr />

EDIT: Adding todo:

- [x] Ensure Dialogs are never larger than the browser window
- [x] When Dialog content is taller than browser window, make it scroll
- [x] Don't allow scrolling on the 'background content' when Dialogs are open
- [x] If the dialog has action buttons, fix them in position so they're always visible if/when the content scrolls
- [x] Remove 'color' declaration currently applied to all Dialog content
- [x] Ensure media Dialog in unaffected by any of these changes

### CODE REVIEW

**Comment by folletto** [comment]
I think the rule of thumb is: a dialog should be simple and small. If there's something with that much data, I'd probably ask: "Can it be a page?". And so far, the answer has always been "yes it can". 

Incidentally, it's also beneficial as a separate page has a position in the IA hierarchy, as well as it becomes linkable directly.

In the example you show above, it seems to me it should be a drill down in a separate page, with the `HeaderCake` component. Why not? :)

**Comment by jameskoster** [comment]
Generally I agree. I also am totally on board with developing standards that third-party authors should follow. But I don't think that's an argument against implementing this as a fallback at the very least.

The gif I shared might not be the best example. A better one would be editing the locations a Shipping Zone covers. An additional screen in that context is dis-orientating as you're already 3 screens deep. Settings > Shipping > Edit a zone > Manage locations. Using a modal there feels more focused.

I'm not convinced we should force plugin authors to use new screens when modals could very well be more appropriate. In those cases I anticipate folks would just create their own scrollable Dialogs anyway which feels like a worse situation to be in.

There is one other caveat; what fits in a simple/small `Dialog` on desktops might not do so on mobile.

**Comment by folletto** [comment]
> An additional screen in that context is dis-orientating as you're already 3 screens deep. Settings > Shipping > Edit a zone > Manage locations. Using a modal there feels more focused.

Note there's no difference between a tap that opens an overlay and a tap that drills down: it's still one tap, it still changes the entire screen.

---

On mobile it should be already handling it, if not, that's a bug :D

**Comment by kellychoffman** [comment]
I think its great to have guidelines around the amount of content within a modal. However, there's no way to know what size screen or what size someone's browser is. Why wouldn't we add something to make sure _everyone_ can view the contents of the modal no matter what?

**Comment by jameskoster** [comment]
> Note there's no difference between a tap that opens an overlay and a tap that drills down: it's still one tap, it still changes the entire screen.

It changes the entire screen visually thanks to the overlay, but you don't _leave_ the screen. So the context remains. You feel grounded in the thing you're editing and the experience is connected.

> On mobile it should be already handling it, if not, that's a bug :D

The issue is the same on mobile although I can't find any dialogs in Calypso core that invoke it. If we fix it for mobile, seems odd to me to not fix it for desktop as well :p

**Comment by folletto** [comment]
> Why wouldn't we add something to make sure everyone can view the contents of the modal no matter what?

I agree. So far didn't emerge as an issue related to scrollbars afaik, that's why it didn't happen. ¯\_(ツ)_/¯ 

> If we fix it for mobile, seems odd to me to not fix it for desktop as well

Yes absolutely. I just thought it was already scrolling on mobile. ;)

**Comment by folletto** [comment]
From the [previous discussion](https://github.com/Automattic/wp-calypso/pull/1293#issuecomment-163169329) for reference:

> Our thinking here before was that dialog should be like browser/OS dialogs, a simple window with a couple options. Once we use it like a full page modal it starts to break, and I'd rather people take conscious measures when they are breaking the pattern (like media).

And even there of course there were two different positions. 

I personally appreciate the argument that if a UI _really_ need to be a dialog, and _really_ needs to scroll, should be handled on a per-dialog basis as it's not how we should be using the dialog. It's not like it's forbidden: it's just not provided.

**Comment by drw158** [comment]
The bottom actions should always be visible.

> A separate but related issue I'm keen to address; it bugs me that when a dialog is scrollable the content behind it can be scrolled too.

We come across this with notifications, and while I think it's probably hard to do, it would be ideal if the content behind didn't scroll.

**Comment by jameskoster** [comment]
> should be handled on a per-dialog basis

The concern there is inconsistency in approach. If we want Calypso to be a development framework I feel like this is definitely going to come up down the road. I know it's unlikely to cause the sky to fall down, but, well, I'm just not sure why we should make folks worry about this when we could just take care of it for them.

Allowing modals to scroll isn't _encouraging_ the behaviour. It's just a fallback that also covers the small-screen issue.

**Comment by folletto** [comment]
> The bottom actions should always be visible.

+1

> Allowing modals to scroll isn't encouraging the behaviour. It's just a fallback that also covers the small-screen issue.

I disagree, but not enough to push back too much on this. ;)

---

> it would be ideal if the content behind didn't scroll.

We have some work going on on the custom scrollbars that might help on that. #14996  It might be a follow-up PR from that one tho.

**Comment by jameskoster** [comment]
> it would be ideal if the content behind didn't scroll

For `Dialog`s this is actually a trivial change thanks to the class that gets added to `body` when `Dialog`s are open/closed :)

**Comment by kellychoffman** [comment]
> it would be ideal if the content behind didn't scroll

That feels like a whole 'nother issue that should be addressed separately. Let's first get the modal itself scrollable so people can always read the content

**Comment by jameskoster** [comment]
> This shouldn't be too difficult, I'll work on it Monday.

Ok, this was trickier than I expected xD

Positioning the elements as we do in the media modal won't work because we don't know the height of the Dialog(s). I've got something working which involves some creative use of min/max-heights. But it doesn't feel great and will require touching the media modal which I'd rather avoid. Not sure if you had an idea of how to approach this @drw158 ?

Here's what I have:

scrolly

**Comment by mladenplavsic** [comment]
How about something like this?

Max height is 100%, so if your screen is small - we use it all, no need for top and bottom margins. Action buttons are always visible. And there is also a 24px gradient (from transparent to white) at the bottom of the content which tells user that there is something bellow (if scrollbar is not visible).

ezgif-3-db7e24ed12

**Comment by jameskoster** [comment]
That looks really good. Wanna open a PR? :)

**Comment by jameskoster** [comment]
I wasn't able to get the buttons to stick in place without applying a height to the dialog (like we do for the media modal). @mladenplavsic if you were able to do that please share how xD

The height/width is applied based on the `top`, `left`, `right` and `bottom` values on the dialog itself. So easily changed for specific dialogs as required.

Here's how a scrollable dialog with action buttons looks on mobile;

mobile

And desktop:

desktop

The drawback here is that now the dialog has a height relative to the browser window, it can result in a lot of whitespace:

whitespace

To get around this, one simply needs to tweak the `top`, `left`, `right` and `bottom` values as required. Given this whitespace is already an 'issue' for the media inserter, hopefully it's not a huge deal. Very keen to hear if anyone has a better / alternative approach to this though.

**Comment by mladenplavsic** [comment]
@jameskoster I created a PR #15789 (as you suggested). Have you tried it? Or you meant something else?

I've done it by using ```flex``` and ```max-height``` on ```.dialog```, among other changes.

**Comment by jameskoster** [comment]
Added a 'title' prop so that Dialogs can optionally be passed a title to display on the frontend like so;

title

Some times it's useful for these Dialogs to have a title, a consistent way to add/display them makes sense I think.

Currently it's not fixed in position (like the buttons). I'm not sure if it needs to be.

**Comment by jameskoster** [comment]
@mladenplavsic sorry, I didn't get a notification for that for some reason o_O.

Sorry to be a pain, but would you mind sending the PR to this branch?

**Comment by folletto** [comment]
> Some times it's useful for these Dialogs to have a title, a consistent way to add/display them makes sense I think.

It's again related to the original idea of a dialog to be minimal... may I suggest at this point given all the suggestions that it probably make sense to create a new component called like `RichDialog` that build upon this and adds all the features you need standardizing? :)

**Comment by jameskoster** [comment]
> It's again related to the original idea of a dialog to be minimal... may I suggest at this point given all the suggestions that it probably make sense to create a new component called like RichDialog that build upon this and adds all the features you need standardizing? :)

I'll think I'll just remove this for now. We don't need it specifically. I'll open a separate PR to discuss if required :)

> I would go with max-height: 100% - why taking space around .dialog on small screens?

Purely a visual thing, it feels clearer to me that it's a modal with a little space around all edges. Here's how it looks:

edit-address

**Comment by jameskoster** [comment]
Keyboard shortcuts was a Dialog that had specific scrolling styles. Removed those and it's all good now. Please check again at your leisure @drw158 :)

**Comment by folletto** [comment]
Did a round of testing, seems behaving well 👍 

screen shot 2017-07-06 at 12 02 42

**Comment by folletto** [comment]
I noticed... The categories dialog is a bit off... in production too:

screen shot 2017-07-06 at 12 11 35

Which with this PR becomes odder:

screen shot 2017-07-06 at 12 12 09

But the outer scrollbar seems already present in production, so it's not a regression from this PR. The second inner scrollbar is new, and without being its fault adds to the awkwardness.

I guess it's again of an instance where things have been developed with invisible scrollbars ^^'

**Comment by jameskoster** [comment]
Fixed category Dialog;

categories

**Comment by folletto** [comment]
Confirmed with scrollbars on 👍 

screen shot 2017-07-06 at 16 46 12

**Comment by drw158** [comment]
Found a weird one after activating a theme. Otherwise looks good to me

download-1

Notice how the buttons are flowing out of the container

**Comment by jameskoster** [comment]
@drw158 @folletto Are you folks able to review this one for me again? Or suggest someone else to review? :)

**Comment by folletto** [comment]
Hi! I don't get notifications for commits, so I had no idea this was updated until now. Checking :)

**Comment by folletto** [comment]
This seems behaving properly now:

screen shot 2017-07-24 at 12 21 48

However, I also noticed that now the dialog lost its default min height, so now the loader starts small and then resizes, which isn't the intended behaviour.

Current PR:

screen shot 2017-07-24 at 12 22 36

Live version:

screen shot 2017-07-24 at 12 23 39


Is there any way to preserve the default size the dialog had? Or we have to check the dialog heights (probably just that one)?

**Comment by jameskoster** [comment]
@folletto thanks and done!

File: package.json [hunk 1]
```diff
@@ -121,8 +121,8 @@
     "react-day-picker": "6.0.2",
     "react-docgen": "2.13.0",
     "react-dom": "15.4.0",
+    "react-modal": "2.2.1",
```
Comment (allendav) [inline package.json]: Recommend maintaining alphabetical order here

**Pull request merged**